### PR TITLE
[TokenAnalysis][CreateLocks] Support ShimDMAOp

### PR DESCRIPTION
**Pull request by hanchenye**

### CODE REVIEW

File: lib/AIECreateLocks.cpp [hunk 1]
```diff
@@ -51,9 +55,12 @@ struct Token2LockLowering : public OpConversionPattern<UseTokenOp> {
...
       srcRow = mem.rowIndex();
       IsParentMemOp = true;
+    } else if (auto shimDma = dyn_cast<ShimDMAOp>(parentOp)) {
+      srcCol = shimDma.colIndex();
+      srcRow = shimDma.rowIndex();
     } else {
-      llvm_unreachable(
-          "A parent operation of UseTokenOp must be either CoreOp or MemOp");
+      llvm_unreachable("A parent operation of UseTokenOp must be either CoreOp "
+                       "or MemOp or ShimDMAOp");
```
Comment (stephenneuendorffer) [inline lib/AIECreateLocks.cpp:63]: probably this should be moved to a verifier?  Actually I was thinking we should extract all of the 'token' stuff into a separate dialect entirely.

**Review by stephenneuendorffer — CHANGES_REQUESTED** [review]
Add a test please?

File: lib/AIECreateLocks.cpp (hunk 1 above)
Comment (hanchenye) [reply lib/AIECreateLocks.cpp:63]: Added a verifier for UseTokenOp.

File: test/create-locks/test_lock_shimdma.mlir [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+module @test_lock_shimdma {
+  AIE.token(0) {sym_name = "token0"}
+  %buf_ext = AIE.external_buffer 0 : memref<256xi32>
+
+  %t60 = AIE.tile(6, 0)
+  %c60 = AIE.core(%t60) {
+    AIE.useToken @token0(Acquire, 2)
+    AIE.useToken @token0(Release, 3)
+    AIE.end
+  }
```
Comment (hanchenye) [inline test/create-locks/test_lock_shimdma.mlir:67]: Shim tile should not have a CoreOp in theory. The token uses here should be located in a runtime dialect operation. I'm experimenting on a runtime dialect locally and will file a PR when it is more mature.

Comment (hanchenye) [reply test/create-locks/test_lock_shimdma.mlir:67]: Based on my experiments, a CoreOp associated with a shim tile will not break the aiecc.

Comment (stephenneuendorffer) [reply test/create-locks/test_lock_shimdma.mlir:67]: I don't understand the intention here... can you elaborate?

Comment (hanchenye) [reply test/create-locks/test_lock_shimdma.mlir:67]: The token uses here should happen on the host CPU sides such that we can avoid data pollution & enable double buffering between DDR and host CPU.

For example in this case, only if the host CPU has acquired token0 with value 2, it should begin to use the data on the corresponding external buffer. Similarly, if value 3 is not released by the host CPU, the shim tile should not launch the DMA. Does this make sense?

Comment (stephenneuendorffer) [reply test/create-locks/test_lock_shimdma.mlir:67]: Got it.. one thing I've pondered is having a host CPU exist in the MLIR, which might be a better place for this.. In the meantime, this is probably good enough..  Maybe you can capture this as a TODO in the file.

Comment (hanchenye) [reply test/create-locks/test_lock_shimdma.mlir]: Added a TODO note here!

**Comment by hanchenye** [comment]
Hi @stephenneuendorffer, could you help to merge this PR?

**Comment by hanchenye** [comment]
Thanks!

**Pull request merged**